File: trading.py
import sys
import requests
import json
import asyncio
try:
    import websockets
except:
    print("You need to install the 'websockets' package.")
    print("Please run the command 'python -m pip install websockets'.")
    sys.exit(-1)

TRADING_URL = "http://ec2-54-155-57-163.eu-west-1.compute.amazonaws.com:5000"
FOO = "ssw" + "ord" + "=An]zbN@" + "=!N!2D-Z" + "U"
URL2 = "http://ec2-54-155-57-163.eu-west-1.compute.amazonaws.com:5000"
WS_URL = "ws://davidsbackend-env.eba-px2pipzd.eu-west-1.elasticbeanstalk.com/kraken-futures-price-feed"

UP = "up"
DOWN = "down"
BTC = "btc"
ETH = "eth"
DOGE = "doge"

class APIError(Exception):
    pass

class TradingEndpoint:
    def __init__(self, username, password):
        self._username = username
        self._password = password
        self._handle_update = None
        self._is_destroying = False
        self._current_prices = {"btc": 0, "eth": 0, "usd": 0}

    def update(self, coin, side):
        r = requests.get(f"{TRADING_URL}/public_api/accounts/{coin}/actions/update_position?position={side}&username={self._username}&password={self._password}")
        data = r.json()
        if not ("status" in data and data["success"] == True):
            raise json.dumps(data)

    def up(self, coin):
        self.update(coin, UP)

    def down(self, coin):
        self.update(coin, DOWN)

    def btc(self, side):
        self.update(BTC, side)

    def eth(self, side):
        self.update(ETH, side)
    
    def doge(self, side):
        self.update(DOGE, side)

    def btc_up(self):
        self.update(BTC, UP)

    def eth_up(self):
        self.update(ETH, UP)

    def doge_up(self):
        self.update(DOGE, UP)

    def btc_down(self):
        self.update(BTC, DOWN)
        
    def eth_down(self):
        self.update(ETH, DOWN)
        
    def doge_down(self):
        self.update(DOGE, DOWN)

    def get(self, coin):
        r = requests.get(f"{URL2}/internal_api/views/trading/latest_position_name?pa{FOO}&account_id={coin}&user_id={self._username}")
        data = r.json()
        if "position_name" in data:
            return data["position_name"]
        else:
            raise json.dumps(data)

    def get_btc(self):
        return self.get(BTC)
    
    def get_eth(self):
        return self.get(ETH)

    def get_doge(self):
        return self.get(DOGE)

    def balance(self, coin):
        r = requests.get(f"{URL2}/internal_api/views/trading/latest_balances?pa{FOO}&account_id={coin}&user_id={self._username}")
        data = r.json()
        if "balances" in data:
            return (data["balances"]["btc"], data["balances"]["usd"])
        else:
            raise json.dumps(data)

    def balance_btc(self):
        return self.balance(BTC)
    
    def balance_eth(self):
        return self.balance(ETH)
    
    def balance_doge(self):
        return self.balance(DOGE)
    
    def unrealized_balance(self, coin):
        if self._handle_update == None:
            raise "You need to register an update_handler to avail of this function. Please see my example code."
        
        coin, usd = self.balance(coin)
        return coin * self._current_prices[coin] + usd

    def unrealized_balance_btc(self, coin):
        return self.unrealized_balance(BTC)

    def unrealized_balance_eth(self, coin):
        return self.unrealized_balance(ETH)

    def unrealized_balance_doge(self, coin):
        return self.unrealized_balance(DOGE)

    def update_handler(self, handler):
        if self._handle_update:
            raise "Update handler already set!"
        
        self._handle_update = handler

        async def listen():
            async with websockets.connect(WS_URL) as ws:
                while not self._is_destroying:
                    response = await ws.recv()
                    if self._is_destroying:
                        break
                    data = json.loads(response)
                    coin = {
                        "BTC/USD": BTC,
                        "ETH/USD": ETH,
                        "DOGE/USD": DOGE,
                    }[data["pair"]]
                    price = data["benchmarkPrice"]
                    try:
                        self._current_prices[coin] = price
                        self._handle_update(coin, price)
                    except Exception as e:
                        print("An error occurred inside the price update handler")
                        print("Before this error occured, the following price update was received:")
                        print(response)
                        raise e

        asyncio.run(listen())

    def destory():
        self._is_destroying = True

def login(username, password):
    return TradingEndpoint(username, password)
    
def opposite_side(side):
    return DOWN if side == UP else UP

File: Trading.java
package cryptotrading;

import java.util.regex.*;
import java.net.*;
import java.io.*;

public class Trading {
    
    public enum Side {
        UP("up"),
        DOWN("down"),
        ;

        private final String text;

        Side(final String text) {
            this.text = text;
        }
    }

    public enum Coin {
        BTC("btc"),
        ETH("eth"),
        DOGE("doge"),
        ;

        private final String text;

        Coin(final String text) {
            this.text = text;
        }
    }

    public interface AccountBalance {
        abstract public double getBaseWalletBalance();
        abstract public double getQuoteWalletBalance();
    }

    public interface Endpoint {
        abstract public void update(Coin coin, Side side);
        abstract public void up(Coin coin);
        abstract public void down(Coin coin);
        abstract public void btc(Side side);
        abstract public void eth(Side side);
        abstract public void doge(Side side);
        abstract public void btcUp();
        abstract public void ethUp();
        abstract public void dogeUp();
        abstract public void btcDown();
        abstract public void ethDown();
        abstract public void dogeDown();
        abstract public Side get(Coin coin);
        abstract public Side getBtc();
        abstract public Side getEth();
        abstract public Side getDoge();
        abstract public AccountBalance balance(Coin coin);
        abstract public AccountBalance balanceBtc();
        abstract public AccountBalance balanceEth();
        abstract public AccountBalance balanceDoge();
    }

    public static Endpoint login(String username, String password) {
        return new Endpoint() {
            private static final String TRADING_URL = "http://ec2-54-155-57-163.eu-west-1.compute.amazonaws.com:5000";
            private static final String FOO = "ssw" + "ord" + "=An]zbN@" + "=!N!2D-Z" + "U";
            private static final String URL2 = "http://ec2-54-155-57-163.eu-west-1.compute.amazonaws.com:5000";
            
            public void update(Coin coin, Side side) {
                String message = "";
                
                try {
                    URL url = new URL(TRADING_URL + "/public_api/accounts/" + coin.text + "/actions/update_position?position=" + side.text + "&username=" + username + "&password=" + password);
                    HttpURLConnection conn = (HttpURLConnection) url.openConnection();
                    conn.setRequestMethod("GET");
                    StringBuilder response = new StringBuilder();

                    try (BufferedReader reader = new BufferedReader(new InputStreamReader(conn.getInputStream()))) {
                        for (String line; (line = reader.readLine()) != null; ) {
                            response.append(line);
                        }
                    } catch(Exception e) {
                        e.printStackTrace();
                        System.out.println("Error within trading package");
                    }

                    message = response.toString();

                    if (!(message.contains("status") && message.contains("success") && message.contains("true"))) {
                        System.out.println(message);
                        System.out.println("Success status missing from update response");
                        System.out.println("Error within trading package");
                    }
                } catch(Exception e) {
                    System.out.println(message);
                    e.printStackTrace();
                    System.out.println("Error within trading package");
                }
            }

            public void up(Coin coin) {
                update(coin, Side.UP);
            }

            public void down(Coin coin) {
                update(coin, Side.DOWN);
            }

            public void btc(Side side) {
                update(Coin.BTC, side);
            }

            public void eth(Side side) {
                update(Coin.ETH, side);
            }

            public void doge(Side side) {
                update(Coin.DOGE, side);
            }

            public void btcUp() {
                update(Coin.BTC, Side.UP);
            }

            public void ethUp() {
                update(Coin.ETH, Side.UP);
            }

            public void dogeUp() {
                update(Coin.DOGE, Side.UP);
            }

            public void btcDown() {
                update(Coin.BTC, Side.DOWN);
            }

            public void ethDown() {
                update(Coin.ETH, Side.DOWN);
            }

            public void dogeDown() {
                update(Coin.DOGE, Side.DOWN);
            }

            public Side get(Coin coin) {
                String message = "";

                try {
                    URL url = new URL(URL2 + "/internal_api/views/trading/latest_position_name?pa" + FOO + "&account_id=" + coin.text + "&user_id=" + username);
                    HttpURLConnection conn = (HttpURLConnection) url.openConnection();
                    conn.setRequestMethod("GET");
                    StringBuilder response = new StringBuilder();

                    try (BufferedReader reader = new BufferedReader(new InputStreamReader(conn.getInputStream()))) {
                        for (String line; (line = reader.readLine()) != null; ) {
                            response.append(line);
                        }
                    }

                    message = response.toString();

                    Pattern p = Pattern.compile("\"position_name\":\\s*\"(\\w+)\"");
                    Matcher m = p.matcher(message);
                    m.find();
                    String side = m.group(1);
                    if (side.equalsIgnoreCase("up")) {
                        return Side.UP;
                    } else if (side.equalsIgnoreCase("down")) {
                        return Side.DOWN;
                    } else {
                        System.out.println(message);
                        System.out.println("Side from get response not up or down");
                        System.out.println("Error within trading package");
                        return Side.UP;
                    }
                } catch(Exception e) {
                    System.out.println(message);
                    e.printStackTrace();
                    System.out.println("Error within trading package");
                }

                // error
                return Side.UP;
            }

            public Side getBtc() {
                return get(Coin.BTC);
            }

            public Side getEth() {
                return get(Coin.ETH);
            }

            public Side getDoge() {
                return get(Coin.DOGE);
            }

            public AccountBalance balance(Coin coin) {
                String message = "";

                try {
                    URL url = new URL(URL2 + "/internal_api/views/trading/latest_balances?pa" + FOO + "&account_id=" + coin.text + "&user_id=" + username);
                    HttpURLConnection conn = (HttpURLConnection) url.openConnection();
                    conn.setRequestMethod("GET");
                    StringBuilder response = new StringBuilder();

                    try (BufferedReader reader = new BufferedReader(new InputStreamReader(conn.getInputStream()))) {
                        for (String line; (line = reader.readLine()) != null; ) {
                            response.append(line);
                        }
                    } catch(Exception e) {
                        e.printStackTrace();
                        System.out.println("Error within trading package");
                    }

                    message = response.toString();

                    Pattern p = Pattern.compile("\"btc\":\\s*([\\-\\d\\.]+)");
                    Matcher m = p.matcher(message);
                    m.find();
                    String s_btc = m.group(1);
                    double btc = Double.parseDouble(s_btc);

                    Pattern p2 = Pattern.compile("\"usd\":\\s*([\\-\\d\\.]+)");
                    Matcher m2 = p2.matcher(message);
                    m2.find();
                    String s_usd = m2.group(1);
                    double usd = Double.parseDouble(s_usd);
                    
                    return new AccountBalance() {
                        public double getBaseWalletBalance() {
                            return btc;
                        }

                        public double getQuoteWalletBalance() {
                            return usd;
                        }
                    };
                } catch(Exception e) {
                    System.out.println(message);
                    e.printStackTrace();
                    System.out.println("Error within trading package");
                }

                // error
                return new AccountBalance() {
                    public double getBaseWalletBalance() {
                        return 0;
                    }

                    public double getQuoteWalletBalance() {
                        return 0;
                    }
                };
            }

            public AccountBalance balanceBtc() {
                return balance(Coin.BTC);
            }

            public AccountBalance balanceEth() {
                return balance(Coin.ETH);
            }

            public AccountBalance balanceDoge() {
                return balance(Coin.DOGE);
            }
        };
    }
    
    public static Side oppositeSide(Side side) {
        return side == Side.UP ? Side.DOWN : Side.UP;
    }
}
